### PR TITLE
[PDI-17758] - Thread Deadlock using data services data source from sp…

### PR DESCRIPTION
**Pull request by bantonio82**
…oon, also hammering the server

@pentaho/tatooine @pentaho/kashyyyk

### CODE REVIEW

File: pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/jdbc/RemoteClient.java [hunk 1]
```diff
@@ -264,21 +270,31 @@ public DataInputStream query( String sql, StreamingMode windowMode, long windowS
...
 
-  @Override public List<String> getServiceNames() throws SQLException {
+  private List<String> getServices( String serviceName ) throws SQLException {
     List<String> serviceNames = new ArrayList<String>();
     try {
-      String result = execService( "/listServices" );
+      StringBuilder serviceArguments = new StringBuilder().append( "/listServices" );
+      if ( StringUtils.isNotBlank( serviceName ) ) {
+        serviceArguments.append( "?serviceName=" );
+        serviceArguments.append( serviceName );
```
Comment (JoaoRMF) [inline pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/jdbc/RemoteClient.java]: Do a URL encoding of this value. The service name can, for example, have spaces and/or special characters.

**Review by JoaoRMF — CHANGES_REQUESTED** [review]
Review also, the license headers

File: pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/jdbc/RemoteClient.java [hunk 2]
```diff
@@ -264,21 +270,31 @@ public DataInputStream query( String sql, StreamingMode windowMode, long windowS
...
-    return getServiceNames();
+    return getServices( serviceName );
   }
 
-  @Override public List<String> getServiceNames() throws SQLException {
+  private List<String> getServices( String serviceName ) throws SQLException {
     List<String> serviceNames = new ArrayList<String>();
     try {
-      String result = execService( "/listServices" );
+      StringBuilder serviceArguments = new StringBuilder().append( "/listServices" );
```
Comment (JoaoRMF) [inline pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/jdbc/RemoteClient.java]: Where is this new parameter read? I think I'm missing the PR for this part?

File: pdi-dataservice-client/src/test/java/org/pentaho/di/trans/dataservice/jdbc/RemoteClientTest.java [hunk 3]
```diff
@@ -387,6 +390,41 @@ public void testExecMethod() throws Exception {
     assertThat( remoteClient.execService( "/status" ), equalTo( "kettle status" ) );
   }
 
+  @Test
+  public void testgetServiceNames() throws Exception {
```
Comment (JoaoRMF) [inline pdi-dataservice-client/src/test/java/org/pentaho/di/trans/dataservice/jdbc/RemoteClientTest.java]: Camel case method name is missing

**Review by JoaoRMF — CHANGES_REQUESTED** [review]

**Review by JoaoRMF — APPROVED** [review]

**Review by sailingscally — APPROVED** [review]

**Pull request merged**